### PR TITLE
Render alpha in opacity, reduce CSS shadow strength

### PR DESCRIPTION
**Pull request by CL-Jeremy**
This pull request aims at improving the rendering strategies of ASS.js to better match that of native SSA renderers based on my own understanding (without looking at the sources of ```ASFilter/libass``` and derivatives).

The main reason for these changes was to make my userscript achieve a better result in displaying real-time comment overlay (aka. "danmaku") on an online streaming site. I chose to build the whole workflow on ASS so watchers could cache the video and watch offline with mainstream video players by loading the comments from the ASS file.

The changes in this pull request are the following:

- the alpha value from the text's fill color is rendered as the opacity of the text layer
- the alpha value from the text's border color remains as the alpha element of the text's background
- the alpha value from the text's shadow color is rendered as if overlayed on 100% white, that is, lightened to the required brightness, without blending with the underlying colors from contents in the video (thus remaining "pure")
- the shadow of a text element is cast only 4 times around it, instead of 9, which is too dark for me when compared with results from native renderers

The effect of the current changes could be viewed on http://acg.tv/av64188727 after installing the userscript on https://mikeslab.dix.asia/pubshare/bilinicomment.user.js with your favorite userscript engine (I recommend Violentmonkey). You may swap the line
```diff
 // @version  0.1
 // @include  *://www.bilibili.com/video/av*
 // @require  https://code.jquery.com/jquery-1.12.4.min.js
-// @require  https://mikeslab.dix.asia/pubshare/dist/ass.min.js
+// @require  https://cdn.jsdelivr.net/npm/assjs/dist/ass.min.js
 // @grant    GM_addStyle
 // ==/UserScript==
 
```
to view the example without the above changes.

I have not yet tested these changes thoroughly enough, but hope to shed some light on what may be done to make the result look better. Changes and corrections (especially if I've understood something wrong) are highly welcome. Feel free to reply in Chinese if you find it more comfortable.

### CODE REVIEW

**Comment by CL-Jeremy** [comment]
Sorry for not being able to comply with ```eslint``` twice, but I found it really counter-intuitive to allow nesting templates with ``` `${`${...}`}` ``` in ES6, as I take them as similar to POSIX backticks i. e. like quotes, but they simply aren't designed in that way. It may be better, but confused me.

**Comment by weizhenye** [comment]
>   the alpha value from the text's fill color is rendered as the opacity of the text layer

Yes, fill color (c1, a1 in source code) should effect on text, but your changes set `opacity` with a1, which will effect the whole element, including background, border, shadow. There is an `\alpha` tag in ASS, it's used for setting a1, a2, a3, a4 at once, it uses `opacity` to render.

>   the alpha value from the text's border color remains as the alpha element of the text's background

No, border color (c3, a3) effect only border, not background.

>  the alpha value from the text's shadow color is rendered as if overlayed on 100% white, that is, lightened to the required brightness, without blending with the underlying colors from contents in the video (thus remaining "pure")

You [tint](https://css-tricks.com/snippets/sass/tint-shade-functions/) shadow color (c4, a4), it's not native SSA renderers' behavior in my experience. Following screenshot from Aegisub with libass should reflect the behavior:
image

> the shadow of a text element is cast only 4 times around it, instead of 9, which is too dark for me when compared with results from native renderers

It's 8. Only 4 times will cause lacking of four corners. In your testing ASS file, every dialogue has `\blur`, it blurs four corners so looks fine. Using `text-shadow` to render border and shadow is a bad idea for performance, but it's the most feasible way I can come out.

---

In summary, your changes don't match my understanding of ASS renderers, telling me if I'm misunderstanding. What's more, [animation](https://github.com/weizhenye/ASS/blob/master/src/renderer/animation.js#L143) should also be considered if changing colors and alphas.

But incompatible with native renderer is fact, I guess the main issue is the difference of rendering `\blur`, I have no idea to solve it.

BTW, I recommend you don't use ASS.js to display danmaku. You can take a look at https://github.com/weizhenye/Danmaku , with full custom style and high performance, I think convert origin danmaku to JSON is more convenience.
